lib/init.lua

```diff
@@ -146,8 +146,69 @@ end
 		* is resolved when all input promises resolve
 		* is rejected if ANY input promises reject
 ]]
-function Promise.all(...)
-	error("unimplemented", 2)
+function Promise.all(promises)
+	if type(promises) ~= "table" then
+		error("Please pass an array of promises or values to Promise.all", 2)
+	end
+
+	-- A traceback for extra arguments warning
+	local source = debug.traceback()
+
+	-- If there are no values then return an already resolved promise.
+	if #promises == 0 then
+		return Promise.resolve({})
+	end
+
+	return Promise.new(function(resolve, reject)
+		-- An array to contain our resolved values from the given promises.
+		local resolvedValues = {}
+		-- Keep a count of resolved promises because just checking the resolved values length
+		-- wouldn't account for promises that resolve with nil.
+		local resolvedCount = 0
+		local rejected = false
+
+		-- Called when a single value is resolved and resolves if all are done.
+		local function resolveOne(i, ...)
+			if rejected == true then
+				-- Bail out if this promise has already been rejected.
+				return
+			end
+
+			if select("#", ...) > 1 then
+				local message = (
+					"When using Promise.all, extra return values are " ..
+					"discarded! See:\n\n%s"
+				):format(
+					source
+				)
+				warn(message)
+			end
+
+			resolvedValues[i] = ...
+			resolvedCount = resolvedCount + 1
+
+			if resolvedCount == #promises then
+				resolve(resolvedValues)
+			end
+		end
+
+		for i, value in ipairs(promises) do
+			-- Pass value through if it's not actually a promise.
+			if Promise.is(value) then
+				value:andThen(function(...)
+					resolveOne(i, ...)
+				end)
+				:catch(function(...)
+					-- Only reject if this promise is unrejected and unresolved.
+					if rejected == false then
+						reject(...)
```

Comment:
Do we want some sort of identifier for _which_ promise failed? Maybe not? 😄

Reply:
It might be useful to have this information, but JS's Promise.all doesn't provide the developer with such information. I think the idea is that the rejected result should provide the developer with enough information to handle the issue effectively.

I think I'd again defer to just mirror JS's implementation here. If the developer wants to know what promise specifically rejected, they could also just `:catch` the promise before adding it to the passed array.

Reply:
In this example the resolved array from Promise.all will have the proper value if the promise rejected, or nil if the promise rejected. Additionally, we can handle the promise rejections inside the chained `:catch` block. 

```lua
for i = 1, 10 do
	promises[i] = Promise.new(function(resolve, reject)
		spawn(function()
			wait(i * 0.1)
			if i == 5 then
				return reject("I hate the number 5")
			end
			resolve(i * 2)
		end)
	end):catch(function(err) -- Chain onto promise here and catch errors
		warn(err)
		-- return 99 -- Optionally return a default value to be in the resolved array rather than nil
	end)
end

Promise.all(promises)
	:andThen(function(values)
		for i = 1, #values do
			print(i, values[i]) -- Index 5 is nil because the promise rejected
		end
	end)
	:catch(function(err) -- Never called
		print(err)
	end)
```